### PR TITLE
DNS cleanup: Move more C pieces to Rust - v5

### PR DESCRIPTION
**Pull request by jasonish**
Moves more of the remaining C parser bits to Rust. Most of the C
was just wrappers around Rust functions anyways.

Also a small DHCP cleanup where it also just wrapped Rust.

Previous PR: https://github.com/OISF/suricata/pull/4589

Changes from last PR:
- Introduce silent/ok errors for rule loading errors that
  should still allow -T to pass.
- Return silent/ok when an app-layer event code is unknown
  unless strict-rule-keywords is on, then it will return
  a normal error and -T will fail.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/443
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/800

### CODE REVIEW

**Comment by jasonish** [comment]
Github appears to be having issue with checks being triggered right now.

File: rust/src/dns/dns.rs [hunk 1]
```diff
@@ -821,9 +917,16 @@ pub extern "C" fn rs_dns_tx_get_query_rrtype(tx: &mut DNSTransaction,
...
+    _flow: *const core::Flow,
+    _dir: u8,
+    input: *const u8,
+    len: u32,
+    rdir: *mut u8,
+) -> AppProto {
+    if len == 0 || len < std::mem::size_of::<DNSHeader>() as u32 {
+        return core::ALPROTO_UNKNOWN;
+    }
     let slice: &[u8] = unsafe {
```
Comment (catenacyber) [inline rust/src/dns/dns.rs]: we can use the `build_slice!` macro here.

File: rust/src/dns/dns.rs [hunk 2]
```diff
@@ -821,9 +917,16 @@ pub extern "C" fn rs_dns_tx_get_query_rrtype(tx: &mut DNSTransaction,
 }
 
 #[no_mangle]
-pub extern "C" fn rs_dns_probe(input: *const u8, len: u32, rdir: *mut u8)
-                               -> u8
-{
+pub extern "C" fn rs_dns_probe(
+    _flow: *const core::Flow,
+    _dir: u8,
```
Comment (catenacyber) [inline rust/src/dns/dns.rs]: This should be used cf https://github.com/OISF/suricata/pull/4058

File: rust/src/dns/dns.rs [hunk 3]
```diff
@@ -861,15 +969,116 @@ pub extern "C" fn rs_dns_probe_tcp(direction: u8,
...
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_dns_udp_register_parser() {
+    let default_port = std::ffi::CString::new("[53]").unwrap();
+    let parser = RustParser{
+        name: b"dns\0".as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_UDP,
+        probe_ts: Some(rs_dns_probe),
+        probe_tc: Some(rs_dns_probe),
```
Comment (catenacyber) [inline rust/src/dns/dns.rs]: We should distinguish both probing parsers

File: rust/src/dns/dns.rs (hunk 1 above)
Comment (jasonish) [reply rust/src/dns/dns.rs]: We could, but then I'd get an error about the function being unsafe.  Maybe could just be marked as a TODO til then.

Comment (catenacyber) [reply rust/src/dns/dns.rs]: ok

Comment (jasonish) [reply rust/src/dns/dns.rs]: Oh, but I suppose I should actually make the function unsafe to support my argument ;)

**Pull request closed without merging**